### DISCUSSION TITLE
How to run the report design in loop

### DISCUSSION THREAD

**Discussion by raviprsd18** [eclipse-birt/birt#1430] [Q&A] [answered]
Hi all,

Would some one help me in resolving the following problem.

I have a report that contains multiple tables, charts, images other controls as well. Which suits well for single user data. How to make this report design run in loop to generate single pdf file for multiple users data.

**Let me explain using simple example.**
Say I have users data base that contains users internet billing details.
Till now one user, one billing details.
Now I need to generated a single report that contains all the users billing details in sequence.
So is there any way to loop the report design file controls in loop and execute the report.

**Comment by hvbtup** [comment, marked as answer]
What I would do is: Use an outer dataset for the users, use a list item for it and put the original layout inside the detail area of the list item?
Then say page-break-before: always for the first item inside the list item detail.

**Comment by raviprsd18** [comment]
Hello @hvbargen,

Thanks for the reply. 
I didn't understood the solution provided by you, because I am new to BIRT.

Is it possible to elaborate more on your answer, if possible can you give me some sample code.

**Comment by speckyspooky** [comment]
Hello @raviprsd18 ,

the solution at the end is little bit dependend from your data and from your reporting-requirements.
So I created a small test report based on the classsic-car-model of BIRT.

The report includes a:
- Standard dynamic table
- Grouped Table: Customer-Order-Table  //  1 Table with Table-Groupung ( all data joined at SQL side, my favorit)
- Looped Table: Customer-Order-Table  //  2 Tables, Loop-table and Inner-Table (like mentioned on the comment above, but I use a table instead of list element)

### Screen of the output
![2023-09-22 19_37_06-BIRT Report Viewer – Mozilla Firefox](https://github.com/eclipse-birt/birt/assets/41593722/c4be2c17-30c9-4413-a131-0d356c83bd05)


### Demo report & Demo pdf
[demo-group-loop-report.zip](https://github.com/eclipse-birt/birt/files/12703203/demo-group-loop-report.zip)

[demo-group-loop-report.pdf](https://github.com/eclipse-birt/birt/files/12703206/demo-group-loop-report.pdf)

**Reply by raviprsd18** [reply]
@speckyspooky Thanks for the example.

Let me try now from my side, and will keep you update.

**Reply by raviprsd18** [reply]
@speckyspooky, 

With the screenshot and the pdf attached I understand the example is given for tables. But in my case, my report design have lot of charts, tables, labels, text etc.

In run time how will I assign the data to these controls dynamically ?
I am attaching sample generated report image here for your reference. Here you can see one table and one chart now I need these 2 controls render more number of times based on the loop.


![image](https://github.com/eclipse-birt/birt/assets/46317080/d3184b52-b4c2-4598-9d57-bd3bc6c2ad90)

**Reply by speckyspooky** [reply]
Your loop-dataset will have a set of control-ID which will be the primary key of your diagram-dataset.
On your diagrm-dataset you will exchange this ID from the loop like a parameter inot the diagrmm-dataset and so you have the ID like a filter into your diagram.
You cann add multiple diagrams and tables into your loop table.

I use this mechanism to create reports with the same diagrams & tables in a loop of countries.
Each country will have the same statistics and the loop defines the countries.

**Comment by raviprsd18** [comment]
@speckyspooky 

I am facing the below issue, when I try to open your rptdesign file

![image](https://github.com/eclipse-birt/birt/assets/46317080/84ad4c7b-20cc-4272-b4a8-9cf82faf748e)

**Reply by speckyspooky** [reply]
You can open the report with an text editor and change the version number to 3.2.23.
Afterwards you can use the report with an older BIRT version too.
